_posts/2024-02-19-jco-1.0.md

```diff
@@ -0,0 +1,270 @@
+toolchain built for WASI 0.1. This work predates Wasm Components.
+- [ComponentizeJS](https://github.com/bytecodealliance/ComponentizeJS): A
+JavaScript → WASI toolchain built for WASI 0.2 and Wasm Components. This project
+is newer, and less mature than Javy.
+- [Jco](https://github.com/bytecodealliance/jco): A comprehensive WASI compiler
+toolchain and runtime for JavaScript. This includes a WASI → JS toolchain built
+for WASI 0.2 and Wasm Components.
+
+The goal is for Jco to be the only toolchain you need when working with
+JavaScript/TypeScript and WASI. And in the future we intend to stabilize the
+`jco componentize` subcommand which uses the ComponentizeJS project internally.
+
+## Example
+
+To give you a taste of what it's like to use jco, let’s compile a little Rust
+program to a WASI 0.2.0 component, install jco for Node.js, and then embed the
+newly built component inside of the runtime. If you want to skip ahead to the
+final result, you can [view it here](https://github.com/yoshuawuyts/playground-jco).
+
+### Dependencies and Repository
+
+Because we're going to be writing Rust code, and using it from Node.js we expect
+you to have both a working [Node.js environment](https://github.com/nvm-sh/nvm)
+and [Rust toolchain](https://rustup.rs) installed. Once you have that, we can
+install [`cargo-component`](https://github.com/bytecodealliance/cargo-component)
+to build our Rust code, and [jco] to build it for Node.js.
+
+```bash
+$ npm install -g @bytecodealliance/jco        # Install the jco cli tool
+$ npm install @bytecodealliance/preview2-shim # Install the jco WASI runtime as a local dep
+$ cargo install cargo-component               # Install the `cargo component` subcommand
+```
+
+We can then proceed to create a new project repository, which has both Rust and JS code:
+
+```sh
+$ cargo component new playground-jco --lib # Create a new Rust component
+$ cd playground-jco                        # Enter the newly created project
+$ npm init -y                              # Setup a Node.js environment
+$ touch index.js                           # Create a JS entrypoint
+$ echo node_modules >> .gitignore          # Add node_modules to our .gitignore file
+```
+
+This should leave you with a structure that roughly looks like this:
+
+```bash
+src/lib.rs      # Our Rust library entry point
+wit/world.wit   # Contains our WIT IDL definitions
+Cargo.toml      # Configures our Rust project
+index.js        # Our JS binary entry point
+package.json    # Configures our JS Project
+```
+
+And with that out of the way, we're ready to begin writing some code!
+
+### Defining the WIT interface
+
+There are three steps to this project: defining the WIT interface, writing the
+Rust code to be exported, and writing the JS code to be imported. We'll be
+starting with the WIT interface. For those of you new to Wasm Components: WIT is
+an Interface Definition Language (IDL) which allows you to describe programming
```

Comment:
_define_? describe isn't wrong, of course..... standard header/class language has _define_ for the interface or structural definition, and _implementation_ for the thing that makes it go. :-) Your post, your choice.

Reply:
wikipedia says "describe", so I'm not going to argue. ty for catching that!